compiler/rustc_interface/src/passes.rs

```diff
@@ -640,22 +643,49 @@ pub static DEFAULT_QUERY_PROVIDERS: LazyLock<Providers> = LazyLock::new(|| {
     *providers
 });
 
-pub fn create_global_ctxt<'tcx>(
+pub(crate) fn create_global_ctxt<'tcx>(
     compiler: &'tcx Compiler,
-    crate_types: Vec<CrateType>,
-    stable_crate_id: StableCrateId,
-    dep_graph: DepGraph,
-    untracked: Untracked,
+    mut krate: rustc_ast::Crate,
     gcx_cell: &'tcx OnceLock<GlobalCtxt<'tcx>>,
     arena: &'tcx WorkerLocal<Arena<'tcx>>,
     hir_arena: &'tcx WorkerLocal<rustc_hir::Arena<'tcx>>,
-) -> &'tcx GlobalCtxt<'tcx> {
+) -> Result<&'tcx GlobalCtxt<'tcx>> {
+    let sess = &compiler.sess;
+
+    rustc_builtin_macros::cmdline_attrs::inject(
+        &mut krate,
+        &sess.psess,
+        &sess.opts.unstable_opts.crate_attr,
+    );
+
+    // FIXME modify krate.attrs in place once #65860 is a hard error.
```

Comment:
I disagree with this and would like to see everything expanded as a part of the regular expansion phase (that's why I introduced the current setup in https://github.com/rust-lang/rust/pull/108221), that will allow to expand crate-level macro attributes correctly.
If some crate attribute has to be used before that, then it is an exception that should be clearly visible, such attributes need to have some restrictions (e.g. cannot be macro-generated).

Reply:
Reverted